src/lib/concurrency/transaction_context.cpp

```diff
@@ -26,8 +52,109 @@ CommitID TransactionContext::commit_id() const {
+
+  if (!success) return false;
+
+  for (const auto& op : _rw_operators) {
+    op->rollback_records();
+  }
+
+  _mark_as_rolled_back();
+
+  return true;
+}
+
+bool TransactionContext::commit(std::function<void(TransactionID)> callback) {
+  const auto success = _prepare_commit();
+
+  if (!success) return false;
+
+  for (const auto& op : _rw_operators) {
+    op->commit_records(commit_id());
+  }
+
+  _mark_as_pending_and_try_commit(callback);
+
+  return true;
+}
+
+bool TransactionContext::_abort() {
+  const auto from_phase = TransactionPhase::Active;
+  const auto to_phase = TransactionPhase::Aborted;
+  const auto end_phase = TransactionPhase::RolledBack;
+  auto success = _transition(from_phase, to_phase, end_phase);
+
+  if (!success) return false;
+
+  _wait_for_active_operators_to_finish();
+  return true;
+}
+
+void TransactionContext::_mark_as_rolled_back() {
+  DebugAssert(([this]() {
+                for (const auto& op : _rw_operators) {
+                  if (op->state() != ReadWriteOperatorState::RolledBack) return false;
+                }
+                return true;
+              }()),
+              "All read/write operators need to have been rolled back.");
+
+  _phase = TransactionPhase::RolledBack;
+}
+
+bool TransactionContext::_prepare_commit() {
+  const auto from_phase = TransactionPhase::Active;
+  const auto to_phase = TransactionPhase::Committing;
+  const auto end_phase = TransactionPhase::Committed;
+  const auto success = _transition(from_phase, to_phase, end_phase);
+
+  if (!success) return false;
+
+  DebugAssert(([this]() {
+                for (const auto& op : _rw_operators) {
+                  if (op->state() != ReadWriteOperatorState::Executed) return false;
+                }
+                return true;
+              }()),
+              "All read/write operators need to be in state Executed (especially not Failed).");
+
+  _wait_for_active_operators_to_finish();
+
+  _commit_context = TransactionManager::get()._new_commit_context();
+  return true;
+}
+
+void TransactionContext::_mark_as_pending_and_try_commit(std::function<void(TransactionID)> callback) {
+  DebugAssert(([this]() {
+                for (const auto& op : _rw_operators) {
+                  if (op->state() != ReadWriteOperatorState::Committed) return false;
+                }
+                return true;
+              }()),
+              "All read/write operators need to have been committed.");
+
+  auto context_weak_ptr = std::weak_ptr<TransactionContext>{this->shared_from_this()};
+  _commit_context->make_pending(_transaction_id, [context_weak_ptr, callback](auto transaction_id) {
+    // If the transaction context still exists, set its phase to Committed.
+    if (auto context_ptr = context_weak_ptr.lock()) {
+      context_ptr->_phase = TransactionPhase::Committed;
+    }
+
+    if (callback) callback(transaction_id);
+  });
+
+  TransactionManager::get()._increment_last_commit_id(_commit_context);
```

Comment:
This is probably bad naming that led to the problem above. This only increments the last_commit_id if there are no previous commits that are still pending.